### PR TITLE
Switch to `EarlyBinder` for `const_param_default` and `impl_trait_ref` queries

### LINKED ISSUE #105779: switch to `EarlyBinder` as default
implement https://github.com/rust-lang/types-team/issues/78

> Change all queries which currently have a `bound_X` variant to return `EarlyBinder<T>` by default and remove the `bound_X` version. Add a method `fn EarlyBinder::<T>::subst_identity(self) -> T` if these queries are used in the identity context and don't need to actually substitute anything.
>
> Not having `EarlyBinder` be the default can very easily result in incorrect uses of these queries, e.g. https://github.com/rust-lang/rust/blob/736c675d2ab65bcde6554e1b73340c2dbc27c85a/compiler/rustc_hir_analysis/src/astconv/mod.rs#L2720 which I found while reviewing https://github.com/rust-lang/rust/pull/101947. The `normalize_ty` normalizes `type_of(def_id)` in the wrong environment as we only call `subst` afterwards.

### PR DESCRIPTION
**Pull request by kylematsuda**
Part of the work to close #105779 and implement https://github.com/rust-lang/types-team/issues/78.

Several queries `X` have a `bound_X` variant that wraps the output in `EarlyBinder`. This PR adds `EarlyBinder` to the return type of  `const_param_default` and `impl_trait_ref`, and removes their `bound_X` variants.

r? @lcnr

### CODE REVIEW

**Comment by jackh726** [comment]
We can do this in parts, doesn't have to be a single PR.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 1]
```diff
@@ -1975,14 +1975,14 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
-                let Some(trait_ref) = tcx.impl_trait_ref(impl_def_id) else {
+                let Some(trait_ref) = tcx.bound_impl_trait_ref(impl_def_id) else {
                     // A cycle error occurred, most likely.
                     let guar = tcx.sess.delay_span_bug(span, "expected cycle error");
                     return Err(guar);
                 };
 
                 self.one_bound_for_assoc_type(
-                    || traits::supertraits(tcx, ty::Binder::dummy(trait_ref)),
+                    || traits::supertraits(tcx, ty::Binder::dummy(trait_ref.skip_binder())),
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1985]: this should use `subst_identity`. We're looking at the impl self type while inside of the impl.

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 2]
```diff
@@ -535,12 +535,12 @@ fn check_item_type(tcx: TyCtxt<'_>, id: hir::ItemId) {
...
             };
             debug!("ItemKind::Impl {} with id {:?}", it.ident, it.owner_id);
-            if let Some(impl_trait_ref) = tcx.impl_trait_ref(it.owner_id) {
+            if let Some(impl_trait_ref) = tcx.bound_impl_trait_ref(it.owner_id.to_def_id()) {
                 check_impl_items_against_trait(
                     tcx,
                     it.span,
                     it.owner_id.def_id,
-                    impl_trait_ref,
+                    impl_trait_ref.skip_binder(),
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/check/check.rs:543]: `subst_identity`. We're comparing impl items with trait items *in the context of the impl*, so the impl generic params is correct.

generally, `skip_binder` is pretty much always wrong unless we 1. rebind the binder later on, or 2. only look at something which isn't a `Ty` or the outermost ty constructor or something

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs [hunk 3]
```diff
@@ -180,8 +180,8 @@ fn check_item<'tcx>(tcx: TyCtxt<'tcx>, item: &'tcx hir::Item<'tcx>) {
         // for `T`
         hir::ItemKind::Impl(ref impl_) => {
             let is_auto = tcx
-                .impl_trait_ref(def_id)
-                .map_or(false, |trait_ref| tcx.trait_is_auto(trait_ref.def_id));
+                .bound_impl_trait_ref(def_id.into())
+                .map_or(false, |trait_ref| tcx.trait_is_auto(trait_ref.skip_binder().def_id));
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/check/wfcheck.rs:184]: this `skip_binder` is correct :+1: we're only looking at the `DefId` which does not change when we substitute generic parameters

File: compiler/rustc_hir_analysis/src/coherence/orphan.rs [hunk 4]
```diff
@@ -21,7 +21,7 @@ pub(crate) fn orphan_check_impl(
     tcx: TyCtxt<'_>,
     impl_def_id: LocalDefId,
 ) -> Result<(), ErrorGuaranteed> {
-    let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+    let trait_ref = tcx.bound_impl_trait_ref(impl_def_id.into()).unwrap().skip_binder();
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/coherence/orphan.rs]: needs `subst_identity` as we should looking at the underlying type in `do_orphan_check_impl`

File: compiler/rustc_middle/src/ty/mod.rs [hunk 5]
```diff
@@ -2187,8 +2187,12 @@ impl<'tcx> TyCtxt<'tcx> {
     ) -> Option<ImplOverlapKind> {
         // If either trait impl references an error, they're allowed to overlap,
         // as one of them essentially doesn't exist.
-        if self.impl_trait_ref(def_id1).map_or(false, |tr| tr.references_error())
-            || self.impl_trait_ref(def_id2).map_or(false, |tr| tr.references_error())
+        if self
+            .bound_impl_trait_ref(def_id1)
+            .map_or(false, |tr| tr.skip_binder().references_error())
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/mod.rs]: `subst_identity` here. We check whether the trait ref in it's own context already references an error.

tbh we're actually checking whether the bound trait ref references an error but can't just do `tr.references_error()` because `EarlyBinder` doesn't allow that to prevent users from accidentally doing that.

File: compiler/rustc_monomorphize/src/collector.rs [hunk 6]
```diff
@@ -1351,7 +1351,10 @@ fn create_mono_items_for_default_impls<'tcx>(
                 tcx.def_path_str(item.owner_id.to_def_id())
             );
 
-            if let Some(trait_ref) = tcx.impl_trait_ref(item.owner_id) {
+            if let Some(trait_ref) = tcx
+                .bound_impl_trait_ref(item.owner_id.to_def_id())
+                .map(ty::EarlyBinder::subst_identity)
```
Comment (lcnr) [inline compiler/rustc_monomorphize/src/collector.rs]: think it would be clearer to change this to
```rust
if let Some(trait_ref) = tcx.impl_trait_ref(item.owner_id.to_def_id()) {
    let trait_ref = trait_ref.subst_identity();
}

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 7]
```diff
@@ -77,8 +77,8 @@ pub fn overlapping_impls(
     // a quick check via fast_reject to tell if the impl headers could possibly
     // unify.
     let drcx = DeepRejectCtxt { treat_obligation_params: TreatParams::AsInfer };
-    let impl1_ref = tcx.impl_trait_ref(impl1_def_id);
-    let impl2_ref = tcx.impl_trait_ref(impl2_def_id);
+    let impl1_ref = tcx.bound_impl_trait_ref(impl1_def_id).map(ty::EarlyBinder::subst_identity);
+    let impl2_ref = tcx.bound_impl_trait_ref(impl2_def_id).map(ty::EarlyBinder::subst_identity);
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/coherence.rs]: there we actually pretend to instantiate both sides with inference variables so `skip_binder` is more correct here.

Note that the two impls can have different generic parameters so both can't be in the identity context. the `DeepRejectCtxt` is setup in a way to deal with `ty::Param` which are actually inference variables (but aren't for perf reasons), so it's fine for it to look into types even with `skip_binder`

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 8]
```diff
@@ -461,7 +461,7 @@ pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanChe
 
     // We only except this routine to be invoked on implementations
     // of a trait, not inherent implementations.
-    let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+    let trait_ref = tcx.bound_impl_trait_ref(impl_def_id).unwrap().skip_binder();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/coherence.rs]: that one happens in the identity context, so `subst_identity`

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 9]
```diff
@@ -1911,7 +1911,7 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     return None;
                 }
 
-                let imp = self.tcx.impl_trait_ref(def_id).unwrap();
+                let imp = self.tcx.bound_impl_trait_ref(def_id).unwrap().subst_identity();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs]: actually `skip_binder` as we again pretend like we've instantiated the `ty::Param` with inference vars

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 10]
```diff
@@ -431,7 +431,7 @@ fn report_conflicting_impls<'tcx>(
 pub(crate) fn to_pretty_impl_header(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Option<String> {
     use std::fmt::Write;
 
-    let trait_ref = tcx.impl_trait_ref(impl_def_id)?;
+    let trait_ref = tcx.bound_impl_trait_ref(impl_def_id)?.skip_binder();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs]: `subst_identity`. We want to print the generic parameters here

File: compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs [hunk 11]
```diff
@@ -48,7 +48,7 @@ trait ChildrenExt<'tcx> {
 impl<'tcx> ChildrenExt<'tcx> for Children {
     /// Insert an impl into this set of children without comparing to any existing impls.
     fn insert_blindly(&mut self, tcx: TyCtxt<'tcx>, impl_def_id: DefId) {
-        let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+        let trait_ref = tcx.bound_impl_trait_ref(impl_def_id).unwrap().subst_identity();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs]: `skip_binder`, the specialization graph from the view of users, so they remain bound.

File: compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs [hunk 12]
```diff
@@ -63,7 +63,7 @@ impl<'tcx> ChildrenExt<'tcx> for Children {
     /// an impl with a parent. The impl must be present in the list of
     /// children already.
     fn remove_existing(&mut self, tcx: TyCtxt<'tcx>, impl_def_id: DefId) {
-        let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+        let trait_ref = tcx.bound_impl_trait_ref(impl_def_id).unwrap().subst_identity();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs]: same `skip_binder`

File: compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs [hunk 13]
```diff
@@ -275,7 +275,7 @@ impl<'tcx> GraphExt<'tcx> for Graph {
     ) -> Result<Option<FutureCompatOverlapError<'tcx>>, OverlapError<'tcx>> {
         assert!(impl_def_id.is_local());
 
-        let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+        let trait_ref = tcx.bound_impl_trait_ref(impl_def_id).unwrap().subst_identity();
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs]: :shrug: conceptionally we actually keep the `EarlyBinder` here. So maybe `skip_binder` with a "FIXME: use `EarlyBinder` in `self.children`"?

File: compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs [hunk 14]
```diff
@@ -388,7 +388,7 @@ pub(crate) fn assoc_def(
     impl_def_id: DefId,
     assoc_def_id: DefId,
 ) -> Result<LeafDef, ErrorGuaranteed> {
-    let trait_def_id = tcx.impl_trait_ref(impl_def_id).unwrap().def_id;
+    let trait_def_id = tcx.bound_impl_trait_ref(impl_def_id).unwrap().skip_binder().def_id;
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.trait_id_of_impl

can actually use another method here

File: compiler/rustc_ty_utils/src/implied_bounds.rs [hunk 15]
```diff
@@ -21,14 +21,16 @@ fn assumed_wf_types(tcx: TyCtxt<'_>, def_id: DefId) -> &ty::List<Ty<'_>> {
             assumed_wf_types.extend(liberated_sig.inputs_and_output);
             tcx.intern_type_list(&assumed_wf_types)
         }
-        DefKind::Impl => match tcx.impl_trait_ref(def_id) {
-            Some(trait_ref) => {
-                let types: Vec<_> = trait_ref.substs.types().collect();
-                tcx.intern_type_list(&types)
+        DefKind::Impl => {
+            match tcx.bound_impl_trait_ref(def_id).map(ty::EarlyBinder::subst_identity) {
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/implied_bounds.rs]: this should be `skip_binder` :thinking: ideally `assumed_wf_types` were to also return `EarlyBinder<...>`

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 16]
```diff
@@ -784,9 +784,9 @@ fn trait_impls_for<'a>(
 
         // Look at each trait implementation to see if it's an impl for `did`
         tcx.find_map_relevant_impl(trait_, ty, |impl_| {
-            let trait_ref = tcx.impl_trait_ref(impl_).expect("this is not an inherent impl");
+            let trait_ref = tcx.bound_impl_trait_ref(impl_).expect("this is not an inherent impl");
             // Check if these are the same type.
-            let impl_type = trait_ref.self_ty();
+            let impl_type = trait_ref.subst_identity().self_ty();
```
Comment (lcnr) [inline src/librustdoc/passes/collect_intra_doc_links.rs:789]: `skip_binder`. we actually keep treating the trait ref as if it were bound

**Comment by lcnr** [comment]
But yeah, realized that the change is big enough that doing it for each query (or a few small ones together) separately in different PRs is probably a better idea to both make reviews easier and to prevent merge conflicts.

**Comment by kylematsuda** [comment]
@jackh726 and @lcnr: Thanks, that sounds good. I updated the title and description to reduce the scope of this PR to just the two queries that are already included here.

@lcnr, thanks for the review and the detailed explanations! Tbh I'm not surprised that I got a lot of them wrong, hopefully after studying your explanations I will do slightly better on the next batch :grimacing:

File: compiler/rustc_monomorphize/src/collector.rs (hunk 6 above)
Comment (kylematsuda) [reply compiler/rustc_monomorphize/src/collector.rs]: I found a few other places where I did something similar, so I cleaned those up too

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 17]
```diff
@@ -1555,7 +1555,7 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                 self.tables.impl_defaultness.set(def_id.index, *defaultness);
                 self.tables.constness.set(def_id.index, *constness);
 
-                let trait_ref = self.tcx.impl_trait_ref(def_id);
+                let trait_ref = self.tcx.impl_trait_ref(def_id).map(ty::EarlyBinder::skip_binder);
```
Comment (lcnr) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1558]: also nicer to do that `skip_binder` in the `if let`

File: compiler/rustc_ty_utils/src/ty.rs [hunk 18]
```diff
@@ -360,7 +360,8 @@ fn issue33140_self_ty(tcx: TyCtxt<'_>, def_id: DefId) -> Option<Ty<'_>> {
 
     let trait_ref = tcx
         .impl_trait_ref(def_id)
-        .unwrap_or_else(|| bug!("issue33140_self_ty called on inherent impl {:?}", def_id));
+        .unwrap_or_else(|| bug!("issue33140_self_ty called on inherent impl {:?}", def_id))
+        .skip_binder();
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/ty.rs:364]: `issue33140_self_ty` should be changed to return `Option<EarlyBinder<Ty<'_>>>` as well.

Might be nice to also do that in a separate PR if you're interested.

**Comment by jackh726** [comment]
@bors r=lcnr rollup=never p=1

Bitrotty

**Comment by pnkfelix** [comment]
there is one small regression to bitmaps-full doc, and three improvements to bitmaps-incr-full {check, debug, opt}.

If its not noise, its seems like the right trade off.

Marking as triaged.

@rustbot label: perf-regression-triaged
